src/QsCompiler/CompilerOptimizations/ConstantPropagation.fs

```diff
@@ -27,101 +29,129 @@ type ConstantPropagator(compiledCallables: ImmutableDictionary<QsQualifiedName,
-                | _ ->
-                    let cases = cbList |> Seq.map (fun (c, b) -> (Option.get c, b))
-                    QsConditionalStatement.New (cases, defaultCase) |> QsConditionalStatement
-
-            override statementKind.onForStatement stm =
-                let loopVar = fst stm.LoopItem |> statementKind.onSymbolTuple
-                let iterVals = statementKind.ExpressionTransformation stm.IterationValues
-                let iterValsAsSeq =
-                    match iterVals.Expression with
-                    | RangeLiteral _ -> 
-                        rangeLiteralToSeq iterVals.Expression |> Seq.map (fun x -> wrapExpr (IntLiteral x) Int) |> Some
-                    | ValueArray va -> va :> seq<_> |> Some
-                    | _ -> None
-
-                match iterValsAsSeq with
-                | Some s ->
-                    let iterRange = 
-                        s |> Seq.map (fun x ->
-                            let variableDecl = QsBinding.New ImmutableBinding (loopVar, x) |> QsVariableDeclaration
-                            let variableDeclStatement = {
-                                Statement = variableDecl
-                                SymbolDeclarations = stm.Body.KnownSymbols
-                                Location = Null
-                                Comments = QsComments.New ([], []) }
-                            let innerScope =
-                                { stm.Body with 
-                                    Statements = stm.Body.Statements.Insert(0, variableDeclStatement) }
-                            {   Statement = innerScope |> QsScopeStatement.New |> QsScopeStatement
-                                SymbolDeclarations = stm.Body.KnownSymbols
-                                Location = Null
-                                Comments = QsComments.New ([], []) })
-                    let outerScope = QsScope.New (iterRange, stm.Body.KnownSymbols)
-                    let result = outerScope |> QsScopeStatement.New |> QsScopeStatement |> statementKind.Transform
-                    // printfn "Unrolled for loop!\nOld statement was:\n%O\nNew statement is:\n%O" (printStm 0 (QsForStatement stm)) (printStm 0 result)
-                    result
-                | None ->
-                    let loopVarType = statementKind.TypeTransformation (snd stm.LoopItem)
-                    let body = statementKind.ScopeTransformation stm.Body
-                    QsForStatement.New ((loopVar, loopVarType), iterVals, body) |> QsForStatement
         }
     }
+
+
+/// The StatementKindTransformation used to simplify statements
+and [<AbstractClass>] private StatementOptimizer(stateRef: TransformationState ref) =
+    inherit StatementKindTransformation()
+
+
+    override this.onExpressionStatement ex =
+        let ex = this.ExpressionTransformation ex
+        maybe {
+            let! qualName, newScope = tryInline !stateRef ex
```

Comment:
I think you need to rename this file - the various functionalities are not as modular as they should be.

Reply:
Fixed in future PR